sql: support multiple RLS policies on a table

spilchen opened the pull request:
Previously, row-level security (RLS) tables only ever enforced one policy. This change allows multiple policies to be applied.

When multiple policies exist, their type determines how they are combined:
- Permissive policies are combined using OR
- Restrictive policies are combined using AND
- At least one permissive policy is required; otherwise, the table is treated as deny-all (all rows are filtered out, or all new rows are rejected).

The change affects both the filtering of existing rows (USING expression) and constraints on new rows (WITH CHECK expression).

Epic: CRDB-45203
Release note: none
Informs #136742

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142479)
<!-- Reviewable:end -->

mw5h reviewed (changes requested):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142479)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142479#-OL55uLU6xLHdJHb2SiE:-OL55uLU6xLHdJHb2SiF:b-5q0hu1):*
Is this some sort of standard, or did we come up with this?
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142479#-OL57XYn8qQJ2qAlbsJd:-OL57XYn8qQJ2qAlbsJe:bcvj0qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/808bb809cac87862107c29e1b7b6796136fb4ea4/pkg/sql/opt/optbuilder/row_level_security.go#L60)):*
> ```Go
> 	var combinedExpr tree.TypedExpr
> 	policies := tabMeta.Table.Policies()
> 	for i, group := range [][]cat.Policy{policies.Permissive, policies.Restrictive} {
> ```

nit: this feels a bit forced. I don't see anything really incorrect about it, but I would probably write this block as two calls to a subroutine that iterates over a group with a functor argument for combining (to be clear, I don't think this blocks this pr). 


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142479)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142479#-OL55uLU6xLHdJHb2SiE:-OL5f0bgCzEB6B-xABJA:bw09r2n):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Is this some sort of standard, or did we come up with this?
</blockquote></details>

Yeah, this is how RLS is implemented in postgres and we are just following that for consistency.
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142479#-OL57XYn8qQJ2qAlbsJd:-OL5sHs27eEz-QPrAuyt:bktejrp) ([raw file](https://github.com/cockroachdb/cockroach/blob/808bb809cac87862107c29e1b7b6796136fb4ea4/pkg/sql/opt/optbuilder/row_level_security.go#L60)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

nit: this feels a bit forced. I don't see anything really incorrect about it, but I would probably write this block as two calls to a subroutine that iterates over a group with a functor argument for combining (to be clear, I don't think this blocks this pr).
</blockquote></details>

I removed the nested loop and split it into two separate loops. However, instead of using a subroutine, I ended up creating a closure to handle the core logic of the loop. Let me know your thoughts.


<!-- Sent from Reviewable.io -->

spilchen commented:
@mw5h PTAL

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice!

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142479#-:-OLA9Lbd2bcjXZwgX2mR:b-7yytuz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen committed:
sql: support multiple RLS policies on a table

Previously, row-level security (RLS) tables only ever enforced one
policy. This change allows multiple policies to be applied.

When multiple policies exist, their type determines how they are
combined:
- Permissive policies are combined using OR
matches any permissive policy).
- Restrictive policies are combined using AND
- At least one permissive policy is required; otherwise, the table is
treated as deny-all (all rows are filtered out, or all new rows are
rejected).

The change affects both the filtering of existing rows (USING
expression) and constraints on new rows (WITH CHECK expression).

Epic: CRDB-45203
Release note: none
Informs #136742
